### ISSUE TITLE
Meta-IRQ does not preempt cooperative threads when woken from IRQ  (cortex_m)

### ISSUE THREAD

**Issue by Ronvanl** [zephyrproject-rtos/zephyr#80574]
**Describe the bug**
I believe commit [42036cd](https://github.com/zephyrproject-rtos/zephyr/commit/42036cdbcae270d00b5bd08e805e33bbd3b5787f) broke the ability of Meta-IRQs to preempt all other threads even if those are cooperative. This was working as expected in Zephyr 3.5.0

```
Z_GENERIC_SECTION(.text._HandlerModeExit) void z_arm_exc_exit(void)
{
#ifdef CONFIG_PREEMPT_ENABLED
	/* If thread is preemptible */
	if (_kernel.cpus->current->base.prio >= 0) {
		/* and cached thread is not current thread */
		if (_kernel.ready_q.cache != _kernel.cpus->current) {
			/* trigger a context switch */
			SCB->ICSR |= SCB_ICSR_PENDSVSET_Msk;
		}
	}
#endif /* CONFIG_PREEMPT_ENABLED */
```
`if (_kernel.cpus->current->base.prio >= 0)` prevents the meta-IRQ from being woken when returning from an interrupt unblocking it with e.g sem_give()

The "promise" as stated in the documentation is broken: 
_The thread, once woken, will be guaranteed to run before the current CPU returns into application code._

**To Reproduce**
Have a cooperative thread doing some processing and have a meta-IRQ thread waiting for a semaphore. Unblock the meta-IRQ by calling a sem_give() from IRQ context. The meta-irq will not be woken immediately but I believe at the first possible reschedule point. 

**Expected behavior**
The meta-IRQ should preempt the cooperative thread and after processing return to that cooperative thread.  

**Impact**
(Severe impact) In our case, we are collected data from a front-end in a time-critical manner  (10ms interval) while also running Bluetooth host. Since BT_RX_WQ is a  cooperative thread the meta-IRQ cannot be woken within the time frame especially when connection is being established. 

**Environment :**

 - Zephyr 3.7.0

**Comment by cvinayak** [comment]
@wopu-ot Could someone in your team comment here, I assume Meta-IRQ are used by some of your implementations?

**Comment by wopu-ot** [comment]
Yes, we use Meta-IRQs in our implementation of the controller, but this issue seems to be specific to ARM and therefore does not apply. The test in tests/kernel/sched/metairq tests some of the Meta-IRQ behavior; maybe it could be extended to test for the issue described above.

**Comment by ithinuel** [comment]
Hi, [this](https://github.com/zephyrproject-rtos/zephyr/pull/65071/commits/131ca8a5bee25706926faccb3d126b8cd059d2d5) was introduced as a fix to [#61761](https://github.com/zephyrproject-rtos/zephyr/issues/61761).

From [here](https://docs.zephyrproject.org/latest/kernel/services/threads/index.html#meta-irq-priorities):

> […] there is a special subclass of cooperative priorities at the highest (numerically lowest) end of the priority space: meta-IRQ threads. These are scheduled according to their normal priority, but also have the special ability to preempt all other threads (and other meta-IRQ threads) at lower priorities, even if those threads are cooperative and/or have taken a scheduler lock.

and

> Note that because this breaks the promise made to cooperative threads by the Zephyr API (namely that the OS won’t schedule other thread until the current thread deliberately blocks)

@Ronvanl does replacing `(_kernel.cpus->current->base.prio >= 0)` with the following snippet helps ?
```c
((_kernel.cpus->current->base.prio >= 0) || (_kernel.cpus->current->base.prio < (-CONFIG_NUM_COOP_PRIORITIES)))
```

**Comment by Ronvanl** [comment]
@ithinuel, I replacing the check with the snippet and it doesn't work as it assumes the current thread is a meta-irq thread? Replacing it with the following snippet does work as it checks the cached thread but I am wondering if this check is supposed to be here anyway. I believe the administration is done in `update_cache()`

```
((_kernel.cpus->current->base.prio >= 0) || thread_is_metairq(_kernel.ready_q.cache))
```

Looking at https://github.com/zephyrproject-rtos/zephyr/issues/61761, it doesn't take in account meta-irq threads which according to the documentation (below) can preempt all threads even if they have taken a scheduler lock.

_These are scheduled according to their normal priority, but also have the special ability to preempt all other threads (and other meta-IRQ threads) at lower priorities, even if those threads are cooperative and/or have taken a scheduler lock._

**Comment by Ronvanl** [comment]
Can we confirm this is a regression and if so will it be fixed and backported to 3.7.0 and if so when?

In the meantime, it there a way to overwrite the z_arm_exc_exit fucntion from application code? We are using workspace application where zephyr is cloned from official repo and we do not have our own fork (yet).

**Comment by ithinuel** [comment]
> Can we confirm this is a regression 

I think so yes.

> and if so will it be fixed and backported to 3.7.0

I think it should be backported.

> and if so when?

once a fixed has been proposed & merged.

> In the meantime, is there a way to overwrite the z_arm_exc_exit functionfrom application code? We are using workspace application where zephyr is cloned from official repo and we do not have our own fork (yet).

I don't think so.
Does the snippet you shared in your previous message fix the issue?
If so, I think the fastest would be for you to open a PR with the fix you could validate, ideally with extra tests that would catch future regressions of that feature.

**Comment by Ronvanl** [comment]
The snippet indeed fixes the issue but I think reverting the changes in [42036cd](https://github.com/zephyrproject-rtos/zephyr/commit/42036cdbcae270d00b5bd08e805e33bbd3b5787f) should be the way forward however there is a reason why its added right?

There is a sample which demonstrates the constant latency of the meta-irq threads https://docs.zephyrproject.org/latest/samples/kernel/metairq_dispatch/README.html. Maybe this can be included in the tests with a well-defined criteria

**Comment by ithinuel** [comment]
The reason for this change is that it is a fix to https://github.com/zephyrproject-rtos/zephyr/issues/61761.

I don't know if zephyr is able to validate timing accuracy (this highly depends on the simulator/device).

**Comment by ithinuel** [comment]
@Ronvanl Would you mind creating a PR with that change you could validate or do you prefer me to create it?

**Comment by Ronvanl** [comment]
Would appreciate if you could do it, especially if any (unit) testing is required as I'm not familiar with the test framework.

**Comment by andyross** [comment]
Driving by: scheduler logic really doesn't belong in arch code, and this is a good example why.  It's long since time to migrate cortex-m to arch_switch(), which puts this logic under a z_get_next_switch_handle() API.  I got started in this space a week and a half ago (in the context of "remove PendSV" and not switch/swap per se) and might have something to push soonish.

**Comment by Ronvanl** [comment]
@andyross I see 3.7.1 is released, any chance a fix is included here already?

**Comment by andyross** [comment]
Not yet, I'd expect that work at best to be slow-rolled into the default configs.  For this bug we want to address it in the existing code; the comment above was a design nitpick, basically.

**Comment by kietavainen** [comment]
There is now a test case for this issue in #84509.
